File: WEEK_1/supplementary_material/W1Q2_vary_dt.py
import matplotlib.pyplot as plt   # import package for plot

# Physical parameters
Q_in = 1     #inflow rate
A_t = 1      # tank cross-sectional area (m^2)
A_out = 0.1  # outlet area (m^2)
g = 9.81     # gravity
h0 = 1      # initial water level

# Simulation time parameters
T = 100       # total simulation time 

# outflow rate Q_out
def Q_out(h):
    return A_out * (2 * g * h) ** 0.5

plt.figure(figsize=(10, 5))

# loop over different values for the time step
for dt in [0.01, 0.1, 1, 5, 10, 20]:
    N = int(T / dt)   # number of time steps

    # lists to store the solution 
    times = (N+1)*[0]
    h = (N+1)*[0]
    h[0] = h0
    
    for i in range(N):
        dh = dt * (Q_in - Q_out(h[i]))
        h[i+1] = h[i] + dh 
        times[i+1] = (i+1)*dt
    
    plt.plot(times, h, label = 'dt = ' + str(dt))

plt.xlabel("Time (s)")
plt.ylabel("Water Level (m)")
plt.title("Water Level Over Time for Different dt")
plt.legend()
plt.grid()
plt.show()
    

File: WEEK_1/supplementary_material/W1Q2_vary_dt.m
% Function to define the inflow rate function
Q_in = @(t) 1;
% Function to calculate the outflow rate based on the water level and outlet area
Q_out = @(outlet_area, water_level) sqrt(2*9.81 * max(water_level,0)) * outlet_area;
A_out = 0.1; % Outlet area (m2)
A_t = 1; % Tank cross-sectional area (m2)
h0 = 1; % Initial water level

% Simulation time parameters
T = 100; % Total simulation time (seconds)

% Create figure for plotting
figure()
grid on
hold on

% Loop over different time step sizes
for dt = [0.01, 0.1, 1, 5, 10, 20]
  N = round(T / dt); % Number of time steps
  % Create arrays to store data for plotting
  times = zeros(1, N + 1);
  h = zeros(1, N + 1);
  h(1) = h0;
  t = 0;
  % Simulate the water level over time
  for i = 1:N
      t = t + dt;
      h(i+1) = h(i) + (Q_in(t) - Q_out(A_out, h(i))) * dt;
      times(i+1) = t;
  end
  plot(times, h, '-')
end

legend('dt=0.01', 'dt=0.1', 'dt=1', 'dt=5', 'dt=10', 'dt=20')

